File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-eq-shared-5-4/CalculatedAmount.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CalculatedAmount : ISerialized
    {
        public CalculatedAmount(XmlNode xmlNode)
        {
            XmlNode calculationDatesNode = xmlNode.SelectSingleNode("calculationDates");
            
            if (calculationDatesNode != null)
            {
                if (calculationDatesNode.Attributes["href"] != null || calculationDatesNode.Attributes["id"] != null) 
                {
                    if (calculationDatesNode.Attributes["id"] != null) 
                    {
                        calculationDatesIDRef_ = calculationDatesNode.Attributes["id"].Value;
                        AdjustableRelativeOrPeriodicDates ob = new AdjustableRelativeOrPeriodicDates(calculationDatesNode);
                        IDManager.SetID(calculationDatesIDRef_, ob);
                    }
                    else if (calculationDatesNode.Attributes["href"] != null)
                    {
                        calculationDatesIDRef_ = calculationDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        calculationDates_ = new AdjustableRelativeOrPeriodicDates(calculationDatesNode);
                    }
                }
                else
                {
                    calculationDates_ = new AdjustableRelativeOrPeriodicDates(calculationDatesNode);
                }
            }
            
        
            XmlNode observationStartDateNode = xmlNode.SelectSingleNode("observationStartDate");
            
            if (observationStartDateNode != null)
            {
                if (observationStartDateNode.Attributes["href"] != null || observationStartDateNode.Attributes["id"] != null) 
                {
                    if (observationStartDateNode.Attributes["id"] != null) 
                    {
                        observationStartDateIDRef_ = observationStartDateNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(observationStartDateNode);
                        IDManager.SetID(observationStartDateIDRef_, ob);
                    }
                    else if (observationStartDateNode.Attributes["href"] != null)
                    {
                        observationStartDateIDRef_ = observationStartDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        observationStartDate_ = new AdjustableOrRelativeDate(observationStartDateNode);
                    }
                }
                else
                {
                    observationStartDate_ = new AdjustableOrRelativeDate(observationStartDateNode);
                }
            }
            
        
            XmlNode optionsExchangeDividendsNode = xmlNode.SelectSingleNode("optionsExchangeDividends");
            
            if (optionsExchangeDividendsNode != null)
            {
                if (optionsExchangeDividendsNode.Attributes["href"] != null || optionsExchangeDividendsNode.Attributes["id"] != null) 
                {
                    if (optionsExchangeDividendsNode.Attributes["id"] != null) 
                    {
                        optionsExchangeDividendsIDRef_ = optionsExchangeDividendsNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(optionsExchangeDividendsNode);
                        IDManager.SetID(optionsExchangeDividendsIDRef_, ob);
                    }
                    else if (optionsExchangeDividendsNode.Attributes["href"] != null)
                    {
                        optionsExchangeDividendsIDRef_ = optionsExchangeDividendsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        optionsExchangeDividends_ = new XsdTypeBoolean(optionsExchangeDividendsNode);
                    }
                }
                else
                {
                    optionsExchangeDividends_ = new XsdTypeBoolean(optionsExchangeDividendsNode);
                }
            }
            
        
            XmlNode additionalDividendsNode = xmlNode.SelectSingleNode("additionalDividends");
            
            if (additionalDividendsNode != null)
            {
                if (additionalDividendsNode.Attributes["href"] != null || additionalDividendsNode.Attributes["id"] != null) 
                {
                    if (additionalDividendsNode.Attributes["id"] != null) 
                    {
                        additionalDividendsIDRef_ = additionalDividendsNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(additionalDividendsNode);
                        IDManager.SetID(additionalDividendsIDRef_, ob);
                    }
                    else if (additionalDividendsNode.Attributes["href"] != null)
                    {
                        additionalDividendsIDRef_ = additionalDividendsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        additionalDividends_ = new XsdTypeBoolean(additionalDividendsNode);
                    }
                }
                else
                {
                    additionalDividends_ = new XsdTypeBoolean(additionalDividendsNode);
                }
            }
            
        
            XmlNode allDividendsNode = xmlNode.SelectSingleNode("allDividends");
            
            if (allDividendsNode != null)
            {
                if (allDividendsNode.Attributes["href"] != null || allDividendsNode.Attributes["id"] != null) 
                {
                    if (allDividendsNode.Attributes["id"] != null) 
                    {
                        allDividendsIDRef_ = allDividendsNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(allDividendsNode);
                        IDManager.SetID(allDividendsIDRef_, ob);
                    }
                    else if (allDividendsNode.Attributes["href"] != null)
                    {
                        allDividendsIDRef_ = allDividendsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        allDividends_ = new XsdTypeBoolean(allDividendsNode);
                    }
                }
                else
                {
                    allDividends_ = new XsdTypeBoolean(allDividendsNode);
                }
            }
            
        
        }
        
    
        #region CalculationDates_
        private AdjustableRelativeOrPeriodicDates calculationDates_;
        public AdjustableRelativeOrPeriodicDates CalculationDates_
        {
            get
            {
                if (this.calculationDates_ != null)
                {
                    return this.calculationDates_; 
                }
                else if (this.calculationDatesIDRef_ != null)
                {
                    calculationDates_ = IDManager.getID(calculationDatesIDRef_) as AdjustableRelativeOrPeriodicDates;
                    return this.calculationDates_; 
                }
                else
                {
                      return this.calculationDates_; 
                }
            }
            set
            {
                if (this.calculationDates_ != value)
                {
                    this.calculationDates_ = value;
                }
            }
        }
        #endregion
        
        public string calculationDatesIDRef_ { get; set; }
        #region ObservationStartDate_
        private AdjustableOrRelativeDate observationStartDate_;
        public AdjustableOrRelativeDate ObservationStartDate_
        {
            get
            {
                if (this.observationStartDate_ != null)
                {
                    return this.observationStartDate_; 
                }
                else if (this.observationStartDateIDRef_ != null)
                {
                    observationStartDate_ = IDManager.getID(observationStartDateIDRef_) as AdjustableOrRelativeDate;
                    return this.observationStartDate_; 
                }
                else
                {
                      return this.observationStartDate_; 
                }
            }
            set
            {
                if (this.observationStartDate_ != value)
                {
                    this.observationStartDate_ = value;
                }
            }
        }
        #endregion
        
        public string observationStartDateIDRef_ { get; set; }
        #region OptionsExchangeDividends_
        private XsdTypeBoolean optionsExchangeDividends_;
        public XsdTypeBoolean OptionsExchangeDividends_
        {
            get
            {
                if (this.optionsExchangeDividends_ != null)
                {
                    return this.optionsExchangeDividends_; 
                }
                else if (this.optionsExchangeDividendsIDRef_ != null)
                {
                    optionsExchangeDividends_ = IDManager.getID(optionsExchangeDividendsIDRef_) as XsdTypeBoolean;
                    return this.optionsExchangeDividends_; 
                }
                else
                {
                      return this.optionsExchangeDividends_; 
                }
            }
            set
            {
                if (this.optionsExchangeDividends_ != value)
                {
                    this.optionsExchangeDividends_ = value;
                }
            }
        }
        #endregion
        
        public string optionsExchangeDividendsIDRef_ { get; set; }
        #region AdditionalDividends_
        private XsdTypeBoolean additionalDividends_;
        public XsdTypeBoolean AdditionalDividends_
        {
            get
            {
                if (this.additionalDividends_ != null)
                {
                    return this.additionalDividends_; 
                }
                else if (this.additionalDividendsIDRef_ != null)
                {
                    additionalDividends_ = IDManager.getID(additionalDividendsIDRef_) as XsdTypeBoolean;
                    return this.additionalDividends_; 
                }
                else
                {
                      return this.additionalDividends_; 
                }
            }
            set
            {
                if (this.additionalDividends_ != value)
                {
                    this.additionalDividends_ = value;
                }
            }
        }
        #endregion
        
        public string additionalDividendsIDRef_ { get; set; }
        #region AllDividends_
        private XsdTypeBoolean allDividends_;
        public XsdTypeBoolean AllDividends_
        {
            get
            {
                if (this.allDividends_ != null)
                {
                    return this.allDividends_; 
                }
                else if (this.allDividendsIDRef_ != null)
                {
                    allDividends_ = IDManager.getID(allDividendsIDRef_) as XsdTypeBoolean;
                    return this.allDividends_; 
                }
                else
                {
                      return this.allDividends_; 
                }
            }
            set
            {
                if (this.allDividends_ != value)
                {
                    this.allDividends_ = value;
                }
            }
        }
        #endregion
        
        public string allDividendsIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-eq-shared-5-4/CalculatedAmount.cpp
// CalculatedAmount.cpp 
#include "CalculatedAmount.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CalculatedAmount::CalculatedAmount(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //calculationDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* calculationDatesNode = xmlNode->FirstChildElement("calculationDates");

   if(calculationDatesNode){calculationDatesIsNull_ = false;}
   else{calculationDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- calculationDatesNode , address : " << calculationDatesNode << std::endl;
   #endif
   if(calculationDatesNode)
   {
      if(calculationDatesNode->Attribute("href") || calculationDatesNode->Attribute("id"))
      {
          if(calculationDatesNode->Attribute("id"))
          {
             calculationDatesIDRef_ = calculationDatesNode->Attribute("id");
             calculationDates_ = boost::shared_ptr<AdjustableRelativeOrPeriodicDates>(new AdjustableRelativeOrPeriodicDates(calculationDatesNode));
             calculationDates_->setID(calculationDatesIDRef_);
             IDManager::instance().setID(calculationDatesIDRef_,calculationDates_);
          }
          else if(calculationDatesNode->Attribute("href")) { calculationDatesIDRef_ = calculationDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { calculationDates_ = boost::shared_ptr<AdjustableRelativeOrPeriodicDates>(new AdjustableRelativeOrPeriodicDates(calculationDatesNode));}
   }

   //observationStartDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* observationStartDateNode = xmlNode->FirstChildElement("observationStartDate");

   if(observationStartDateNode){observationStartDateIsNull_ = false;}
   else{observationStartDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- observationStartDateNode , address : " << observationStartDateNode << std::endl;
   #endif
   if(observationStartDateNode)
   {
      if(observationStartDateNode->Attribute("href") || observationStartDateNode->Attribute("id"))
      {
          if(observationStartDateNode->Attribute("id"))
          {
             observationStartDateIDRef_ = observationStartDateNode->Attribute("id");
             observationStartDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(observationStartDateNode));
             observationStartDate_->setID(observationStartDateIDRef_);
             IDManager::instance().setID(observationStartDateIDRef_,observationStartDate_);
          }
          else if(observationStartDateNode->Attribute("href")) { observationStartDateIDRef_ = observationStartDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { observationStartDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(observationStartDateNode));}
   }

   //optionsExchangeDividendsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* optionsExchangeDividendsNode = xmlNode->FirstChildElement("optionsExchangeDividends");

   if(optionsExchangeDividendsNode){optionsExchangeDividendsIsNull_ = false;}
   else{optionsExchangeDividendsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- optionsExchangeDividendsNode , address : " << optionsExchangeDividendsNode << std::endl;
   #endif
   if(optionsExchangeDividendsNode)
   {
      if(optionsExchangeDividendsNode->Attribute("href") || optionsExchangeDividendsNode->Attribute("id"))
      {
          if(optionsExchangeDividendsNode->Attribute("id"))
          {
             optionsExchangeDividendsIDRef_ = optionsExchangeDividendsNode->Attribute("id");
             optionsExchangeDividends_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(optionsExchangeDividendsNode));
             optionsExchangeDividends_->setID(optionsExchangeDividendsIDRef_);
             IDManager::instance().setID(optionsExchangeDividendsIDRef_,optionsExchangeDividends_);
          }
          else if(optionsExchangeDividendsNode->Attribute("href")) { optionsExchangeDividendsIDRef_ = optionsExchangeDividendsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { optionsExchangeDividends_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(optionsExchangeDividendsNode));}
   }

   //additionalDividendsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* additionalDividendsNode = xmlNode->FirstChildElement("additionalDividends");

   if(additionalDividendsNode){additionalDividendsIsNull_ = false;}
   else{additionalDividendsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- additionalDividendsNode , address : " << additionalDividendsNode << std::endl;
   #endif
   if(additionalDividendsNode)
   {
      if(additionalDividendsNode->Attribute("href") || additionalDividendsNode->Attribute("id"))
      {
          if(additionalDividendsNode->Attribute("id"))
          {
             additionalDividendsIDRef_ = additionalDividendsNode->Attribute("id");
             additionalDividends_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(additionalDividendsNode));
             additionalDividends_->setID(additionalDividendsIDRef_);
             IDManager::instance().setID(additionalDividendsIDRef_,additionalDividends_);
          }
          else if(additionalDividendsNode->Attribute("href")) { additionalDividendsIDRef_ = additionalDividendsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { additionalDividends_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(additionalDividendsNode));}
   }

   //allDividendsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* allDividendsNode = xmlNode->FirstChildElement("allDividends");

   if(allDividendsNode){allDividendsIsNull_ = false;}
   else{allDividendsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- allDividendsNode , address : " << allDividendsNode << std::endl;
   #endif
   if(allDividendsNode)
   {
      if(allDividendsNode->Attribute("href") || allDividendsNode->Attribute("id"))
      {
          if(allDividendsNode->Attribute("id"))
          {
             allDividendsIDRef_ = allDividendsNode->Attribute("id");
             allDividends_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(allDividendsNode));
             allDividends_->setID(allDividendsIDRef_);
             IDManager::instance().setID(allDividendsIDRef_,allDividends_);
          }
          else if(allDividendsNode->Attribute("href")) { allDividendsIDRef_ = allDividendsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { allDividends_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(allDividendsNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<AdjustableRelativeOrPeriodicDates> CalculatedAmount::getCalculationDates()
{
   if(!this->calculationDatesIsNull_){
        if(calculationDatesIDRef_ != ""){
             return boost::shared_static_cast<AdjustableRelativeOrPeriodicDates>(IDManager::instance().getID(calculationDatesIDRef_));
        }else{
             return this->calculationDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableRelativeOrPeriodicDates>();
   }
}
boost::shared_ptr<AdjustableOrRelativeDate> CalculatedAmount::getObservationStartDate()
{
   if(!this->observationStartDateIsNull_){
        if(observationStartDateIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(observationStartDateIDRef_));
        }else{
             return this->observationStartDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
boost::shared_ptr<XsdTypeBoolean> CalculatedAmount::getOptionsExchangeDividends()
{
   if(!this->optionsExchangeDividendsIsNull_){
        if(optionsExchangeDividendsIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(optionsExchangeDividendsIDRef_));
        }else{
             return this->optionsExchangeDividends_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
boost::shared_ptr<XsdTypeBoolean> CalculatedAmount::getAdditionalDividends()
{
   if(!this->additionalDividendsIsNull_){
        if(additionalDividendsIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(additionalDividendsIDRef_));
        }else{
             return this->additionalDividends_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
boost::shared_ptr<XsdTypeBoolean> CalculatedAmount::getAllDividends()
{
   if(!this->allDividendsIsNull_){
        if(allDividendsIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(allDividendsIDRef_));
        }else{
             return this->allDividends_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
}

